test failures due to src/ApplicationState.jsx not found

`npm test` results in failure:

```
> sight@1.0.0 test /home/mrobins/git/Electron-Sight
> mocha --require babel-core/register

/home/mrobins/git/Electron-Sight/node_modules/mocha/node_modules/yargs/yargs.js:1163
      else throw err
           ^

Error: Cannot find module '../src/ApplicationState'
Require stack:
- /home/mrobins/git/Electron-Sight/test/test_application_state.js
- /home/mrobins/git/Electron-Sight/node_modules/mocha/lib/mocha.js
- /home/mrobins/git/Electron-Sight/node_modules/mocha/lib/cli/one-and-dones.js
- /home/mrobins/git/Electron-Sight/node_modules/mocha/lib/cli/options.js
- /home/mrobins/git/Electron-Sight/node_modules/mocha/bin/mocha
 ```

--- a/package.json
+++ b/package.json
@@ -66,7 +66,8 @@
     "react-dom": "^16.8.6",
     "react-hot-loader": "^3.1.3",
     "react-redux": "^7.1.3",
-    "redux": "^4.0.4"
+    "redux": "^4.0.4",
+    "sharp": "^0.23.4"
   },
   "devDependencies": {
     "babel-plugin-transform-async-to-generator": "^6.24.1",

--- a/package.json
+++ b/package.json
@@ -59,6 +59,7 @@
     "immutable": "^4.0.0-rc.12",
     "libsvm-js": "^0.2.1",
     "node-addon-api": "^1.3.0",
+    "opencv.js": "^1.2.1",
     "openseadragon": "^2.4.0",
     "react": "^16.8.6",
     "react-dom": "^16.8.6",

--- a/package.json
+++ b/package.json
@@ -55,6 +55,7 @@
     "electron-compile": "^6.4.4",
     "electron-devtools-installer": "^2.2.4",
     "electron-squirrel-startup": "^1.0.0",
+    "electron-store": "^5.1.0",
     "google-palette": "^1.1.0",
     "immutable": "^4.0.0-rc.12",
     "libsvm-js": "^0.2.1",

--- a/src/components/Annotations.jsx
+++ b/src/components/Annotations.jsx
@@ -10,12 +10,19 @@ import {
   RadioGroup,
 } from "@blueprintjs/core";
 
+import Store from 'electron-store'
+const remote = require('electron').remote;
+const app = remote.app;
+
+
+
 
 
 class Annotations extends React.Component {
   constructor(props) {
     super(props)
     this.state = {}
+    this.store = new Store({name: 'sight'});
   }
 
   onClick(data) {
@@ -34,6 +41,31 @@ class Annotations extends React.Component {
     this.setState({
       annotation_active: name,
     });
+  }
+
+  saveAnnotationToJSON() {
+    const selected_name = this.props.annotations.getIn(['current', 'name']);
+    let selected_annotation = this.props.annotations.getIn(['created', selected_name]);
+    console.log(`trying to save {name: ${selected_name}, annotation: ${selected_annotation}}`);
+    if (selected_annotation) {
+      console.log(`save current annotation in ${app.getPath('userData')}`);
+      this.store.set('annotation', selected_annotation);
+    }
+  }
+
+  loadAnnotationFromJSON() {
+    let selected_annotation = this.store.get('annotation');
+    if (selected_annotation) {
+      const name = selected_annotation.name;
+      const polygon = selected_annotation.polygon;
+    
+      console.log('load current annoation');
+      this.props.updateName(name);
+      polygon.map((point) => {
+        this.props.addPoint(new OpenSeadragon.Point(point.x, point.y));
+      });
+      this.props.saveAnnotation();
+    }
   }
 
   annotationToReact({name='', value={}, dashed=false, fill_color="green"}) {

--- a/src/components/ApplicationState.jsx
+++ b/src/components/ApplicationState.jsx
@@ -1,6 +1,8 @@
 import React from 'react';
 import {
   Callout,
+  Label,
+  Checkbox,
   InputGroup,
   Button,
   ButtonGroup,
@@ -311,6 +313,23 @@ PredictMode.prototype.predictPopdown = function(menu) {
         Go...
       </Button>
       </ButtonGroup> 
+      <div>
+        <Label>Display:</Label>
+        <Checkbox checked={menu.props.predictShowCells} label="Post-processed cells" 
+                  onChange={(evt) => {
+                    const value = evt.target.checked;
+                    menu.viewer.predict.redrawOverlays(value, menu.props.predictShowSuperpixels);
+                    menu.props.updatePredictShowCells(value);
+                  }}
+        />
+        <Checkbox checked={menu.props.predictShowSuperpixels} label="Superpixel classification" 
+                  onChange={(evt) => {
+                    const value = evt.target.checked;
+                    menu.viewer.predict.redrawOverlays(menu.props.predictShowCells, value);
+                    menu.props.updatePredictShowSuperpixels(value);
+                  }}
+        />
+      </div>
       </div>
     );
 }

--- a/src/components/Classifiers.jsx
+++ b/src/components/Classifiers.jsx
@@ -155,7 +155,7 @@ class Classifiers extends React.Component {
 
       // create overlay
       const tile_overlay = new TileOverlay(tile, outlabels,
-        outCollectedFeatures);
+        outCollectedFeatures, img_data);
 
       // find superpixel user has selected
       const selected_superpixel = outlabels[selected_tile_index];

--- a/src/components/Classifiers.jsx
+++ b/src/components/Classifiers.jsx
@@ -4,6 +4,9 @@ import palette from 'google-palette';
 import _ from 'underscore';
 import Color from 'color';
 import SVM from 'libsvm-js/asm';
+import Store from 'electron-store'
+const remote = require('electron').remote;
+const app = remote.app;
 
 import {
   H5,
@@ -37,6 +40,7 @@ class Classifiers extends React.Component {
   constructor(props) {
     super(props)
     this.state = {};
+    this.store = new Store({name: 'sight'});
   }
 
   onClick(data) {
@@ -279,6 +283,33 @@ class Classifiers extends React.Component {
   loadClassifier(name) {
     console.log(`loading classifier ${name}`);
     this.props.loadClassifier(this.props.classifiers.getIn(['created', name]));
+  }
+
+  saveClassifierToJSON() {
+    const selected_name = this.props.classifiers.getIn(['current', 'name']);
+    let selected_classifier = this.props.classifiers.getIn(['created', selected_name]);
+    selected_classifier = selected_classifier.set('svm', selected_classifier.get('svm').serializeModel());
+    console.log(`trying to save {name: ${selected_name}, classifier: ${selected_classifier}}`);
+    if (selected_classifier) {
+      console.log(`save current classifier in ${app.getPath('userData')}`);
+      this.store.set('classifier', selected_classifier);
+    }
+  }
+
+  loadClassifierFromJSON() {
+    let selected_classifier = this.store.get('classifier');
+    if (selected_classifier) {
+      selected_classifier.svm = SVM.load(selected_classifier.svm);
+      const name = selected_classifier.name;
+      const min = selected_classifier.feature_min;
+      const max = selected_classifier.feature_max;
+      const score = selected_classifier.score;
+      const svm = selected_classifier.svm;
+    
+      console.log('load current classifier');
+      this.props.updateName(name);
+      this.props.saveClassifier(svm, min, max, score);
+    }
   }
 
   render() {

--- a/src/components/Menu.jsx
+++ b/src/components/Menu.jsx
@@ -174,6 +174,7 @@ class Menu extends React.Component {
     );
 
     let predict = (
+      
       <Button 
             icon="circle"
             active={this.props.mode.predictButtonActive()}
@@ -191,6 +192,34 @@ class Menu extends React.Component {
                   onChange={this.changeHandler("contrast")}
                   value={this.state.contrast} />
     );
+    
+    let save = null;
+    if (this.viewer && this.viewer.classifier) {
+      save = (
+        <Button 
+              onClick={() => {
+                this.viewer.classifier.saveClassifierToJSON();
+                this.viewer.annotations.saveAnnotationToJSON();
+              }}
+        >
+          Save
+        </Button>
+      );
+    }
+    
+    let load = null;
+    if (this.viewer && this.viewer.classifier) {
+      load = (
+        <Button 
+              onClick={() => {
+                this.viewer.classifier.loadClassifierFromJSON();
+                this.viewer.annotations.loadAnnotationFromJSON();
+              }}
+        >
+          Load
+        </Button>
+      );
+    }
 
     return (
       <div>
@@ -207,6 +236,8 @@ class Menu extends React.Component {
         {this.state.brightness_active && brightness_popdown}
         {contrast}
         {this.state.contrast_active && contrast_popdown}
+        {save}
+        {load}
       </ButtonGroup>
     </Card>
     <Viewer 

--- a/src/components/Predict.jsx
+++ b/src/components/Predict.jsx
@@ -72,7 +72,7 @@ function predictTile(tile, img_data, classifier) {
         outLABVariances, outCollectedFeatures
       ] = slic.slic(img_data.data, img_data.width, img_data.height, superpixel_size);
 
-    var tile_overlay = new TileOverlay(tile, outlabels, outCollectedFeatures);
+    var tile_overlay = new TileOverlay(tile, outlabels, outCollectedFeatures, img_data);
 
     // generate features
     const min = classifier.get('feature_min');
@@ -89,6 +89,7 @@ function predictTile(tile, img_data, classifier) {
     classifier.get('svm').predict(features).map((x, idx) => {
       tile_overlay.add_classification(idx, (x > 0 ? 1 : -1));
     });
+    tile_overlay.segment();
     tile_overlay.redraw();
     return tile_overlay;
   }
@@ -107,6 +108,14 @@ class Predict extends React.Component {
           element: tile_overlay.canvas,
           location: tile_overlay.tile.bounds
         });
+      });
+    });
+  }
+
+  redrawOverlays() {
+    this.props.results.map(annotation_result => {
+      annotation_result.map((tile_overlay, id) => {
+        tile_overlay.draw
       });
     });
   }

--- a/src/components/Predict.jsx
+++ b/src/components/Predict.jsx
@@ -64,35 +64,35 @@ function predictAnnotation(openseadragon, classifier, annotation, saveTilePredic
 
 // called when a new tile is loaded
 function predictTile(tile, img_data, classifier) {
-    // SLIC the image data
-    const superpixel_size = classifier.get('superpixel_size');
-    const [
-        outlabels, outLABMeanintensities,
-        outPixelCounts, outseedsXY,
-        outLABVariances, outCollectedFeatures
-      ] = slic.slic(img_data.data, img_data.width, img_data.height, superpixel_size);
+  // SLIC the image data
+  const superpixel_size = classifier.get('superpixel_size');
+  const [
+      outlabels, outLABMeanintensities,
+      outPixelCounts, outseedsXY,
+      outLABVariances, outCollectedFeatures
+    ] = slic.slic(img_data.data, img_data.width, img_data.height, superpixel_size);
 
-    var tile_overlay = new TileOverlay(tile, outlabels, outCollectedFeatures, img_data);
+  var tile_overlay = new TileOverlay(tile, outlabels, outCollectedFeatures, img_data);
 
-    // generate features
-    const min = classifier.get('feature_min');
-    const max = classifier.get('feature_max');
-    const n_superpixels = outPixelCounts.length
-    const features = Array(n_superpixels).fill().map((_, idx) => {
-      // scale features
-      return tile_overlay.generate_data(idx).map(
-                  (x, idx) => ((x - min[idx]) / (max[idx] - min[idx]))
-      );
-    });
+  // generate features
+  const min = classifier.get('feature_min');
+  const max = classifier.get('feature_max');
+  const n_superpixels = outPixelCounts.length
+  const features = Array(n_superpixels).fill().map((_, idx) => {
+    // scale features
+    return tile_overlay.generate_data(idx).map(
+                (x, idx) => ((x - min[idx]) / (max[idx] - min[idx]))
+    );
+  });
 
-    // predict on features using svm
-    classifier.get('svm').predict(features).map((x, idx) => {
-      tile_overlay.add_classification(idx, (x > 0 ? 1 : -1));
-    });
-    tile_overlay.segment();
-    tile_overlay.redraw();
-    return tile_overlay;
-  }
+  // predict on features using svm
+  classifier.get('svm').predict(features).map((x, idx) => {
+    tile_overlay.add_classification(idx, (x > 0 ? 1 : -1));
+  });
+  tile_overlay.segment();
+  tile_overlay.redraw();
+  return tile_overlay;
+}
 
 
 class Predict extends React.Component {
@@ -136,6 +136,7 @@ class Predict extends React.Component {
       );
     });
   }
+
 
   render() {
     return null;

--- a/src/components/Predict.jsx
+++ b/src/components/Predict.jsx
@@ -4,11 +4,11 @@ let slic = require('../../addons/slic/slic');
 const electron = window.require('electron');
 import OpenSeadragon from 'openseadragon';
 import TileOverlay from '../util/TileOverlay'
+import TileCellData from '../util/TileCellData'
 import { create_tile, load_and_process_tile } from '../util/openseadragon.js'
 import { inPolygon } from '../util/geometry.js'
 
-function predictAnnotation(openseadragon, classifier, annotation, saveTilePrediction) {
-  console.log(`predict using ${classifier.get('name')} on ${annotation.get('name')}`);
+function predictAnnotation(openseadragon, classifier, annotation, saveTilePrediction, show_cells, show_superpixels) {
   const tile_source = openseadragon.world.getItemAt(0).source;
   const tiled_image = openseadragon.world.getItemAt(0);
   // get minimum bounding rectangle around polygon
@@ -36,11 +36,22 @@ function predictAnnotation(openseadragon, classifier, annotation, saveTilePredic
     // predict for tile
     const tile_overlay = predictTile(tile, img_data, classifier);
 
-    // add overlay to openseadragon viewer
-    openseadragon.addOverlay({
-      element: tile_overlay.canvas,
-      location: tile_overlay.tile.bounds
-    });
+    // render and add overlay to openseadragon viewer
+    if (show_superpixels) {
+      tile_overlay.superpixel_classification.redraw();
+      openseadragon.addOverlay({
+        element: tile_overlay.superpixel_classification.canvas,
+        location: tile_overlay.tile.bounds
+      });
+    }
+    if (show_cells) {
+      tile_overlay.redraw();
+      openseadragon.addOverlay({
+        element: tile_overlay.canvas,
+        location: tile_overlay.tile.bounds
+      });
+    }
+    
 
     // add tile_overlay to store
     saveTilePrediction(annotation.get('name'), tile_overlay);
@@ -89,9 +100,7 @@ function predictTile(tile, img_data, classifier) {
   classifier.get('svm').predict(features).map((x, idx) => {
     tile_overlay.add_classification(idx, (x > 0 ? 1 : -1));
   });
-  tile_overlay.segment();
-  tile_overlay.redraw();
-  return tile_overlay;
+  return new TileCellData(tile_overlay);
 }
 
 
@@ -101,23 +110,35 @@ class Predict extends React.Component {
     this.state = {};
   }
 
-  startPredict() {
+  redrawOverlays(show_cells, show_superpixels) {
+
+    // clear existing overlays
+    this.props.openseadragon.clearOverlays();
+
+    // redraw everything and add to openseadragon overlays
     this.props.results.map(annotation_result => {
       annotation_result.map((tile_overlay, id) => {
-        this.props.openseadragon.addOverlay({
-          element: tile_overlay.canvas,
-          location: tile_overlay.tile.bounds
-        });
+        if (show_superpixels) {
+          tile_overlay.superpixel_classification.redraw();
+          this.props.openseadragon.addOverlay({
+            element: tile_overlay.superpixel_classification.canvas,
+            location: tile_overlay.tile.bounds
+          });
+        }
+        if (show_cells) {
+          tile_overlay.redraw();
+          this.props.openseadragon.addOverlay({
+            element: tile_overlay.canvas,
+            location: tile_overlay.tile.bounds
+          });
+        }
+        
       });
     });
   }
 
-  redrawOverlays() {
-    this.props.results.map(annotation_result => {
-      annotation_result.map((tile_overlay, id) => {
-        tile_overlay.draw
-      });
-    });
+  startPredict() {
+    this.redrawOverlays(this.props.show_cells, this.props.show_superpixels);
   }
 
   endPredict() {
@@ -125,14 +146,15 @@ class Predict extends React.Component {
   }
 
   onPredict() {
-    console.log('called onPredict');
     this.props.openseadragon.clearOverlays();
     this.props.annotations.map(annotation => {
       predictAnnotation(
         this.props.openseadragon, 
         this.props.classifier, 
         annotation, 
-        this.props.saveTilePrediction
+        this.props.saveTilePrediction,
+        this.props.show_cells,
+        this.props.show_superpixels
       );
     });
   }

--- a/src/containers/Menu.jsx
+++ b/src/containers/Menu.jsx
@@ -15,6 +15,12 @@ import {
   predict,
   openFile,
 } from '../redux/modules/applicationState.js'
+import {
+  updateShowCells as updatePredictShowCells,
+  updateShowSuperpixels as updatePredictShowSuperpixels,
+} from '../redux/modules/predict.js'
+
+
 
 import Menu from '../components/Menu.jsx'
 
@@ -27,6 +33,8 @@ const mapStateToProps = (state, ownProps) => {
     classifierSuperpixelSize: state.classifiers.get('current').get('superpixel_size'),
     classifierZoom: state.classifiers.get('current').get('zoom'),
     classifiers: state.classifiers.get('created'),
+    predictShowCells: state.predict.get('show_cells'),
+    predictShowSuperpixels: state.predict.get('show_superpixels'),
     mode: state.applicationState.mode,
   }
 }
@@ -39,6 +47,8 @@ const mapDispatchToProps = (dispatch) => {
     updateClassifierGamma: (gamma) => { dispatch(updateClassifierGamma(gamma)) },
     updateClassifierName: (name) => { dispatch(updateClassifierName(name)) },
     updateAnnotationName: (name) => { dispatch(updateAnnotationName(name)) },
+    updatePredictShowCells: (value) => { dispatch(updatePredictShowCells(value)) },
+    updatePredictShowSuperpixels: (value) => { dispatch(updatePredictShowSuperpixels(value)) },
     predict: (menu) => { dispatch(predict(menu)) },
     buildClick: (menu) => { dispatch(buildClick(menu)) },
     animClick: (menu) => { dispatch(animClick(menu)) },

--- a/src/containers/Predict.jsx
+++ b/src/containers/Predict.jsx
@@ -8,7 +8,9 @@ const mapStateToProps = (state, ownProps) => {
   return {
     classifier: state.classifiers.getIn(['created', state.classifiers.getIn(['current', 'name'])]),
     annotations: state.annotations.get('created'),
-    results: state.predict.results
+    results: state.predict.get('results'),
+    show_cells: state.predict.get('show_cells'),
+    show_superpixels: state.predict.get('show_superpixels')
   }
 }
 

--- a/src/redux/modules/predict.js
+++ b/src/redux/modules/predict.js
@@ -1,18 +1,36 @@
 const { Map, List } = require("immutable");
 
 const SAVE = 'electron-sight/predict/SAVE'
+const UPDATE_SHOW_CELLS = 'electron-sight/predict/UPDATE_SHOW_CELLS'
+const UPDATE_SHOW_SUPERPIXELS = 'electron-sight/predict/UPDATE_SHOW_SUPERPIXELS'
 
 export function saveTilePrediction(annotation_name, tile_overlay) {
   return { type: SAVE, annotation_name, tile_overlay };
 }
-const initialState = { results: Map() };
+
+export function updateShowCells(value) {
+  return { type: UPDATE_SHOW_CELLS, name:'show_cells', value };
+}
+
+export function updateShowSuperpixels(value) {
+  return { type: UPDATE_SHOW_SUPERPIXELS, name:'show_superpixels', value };
+}
+
+const initialState = Map({ 
+  show_cells: true,
+  show_superpixels: false,
+  results: Map() 
+});
 
 export default function reducer(state = initialState, action = {}) {
   switch (action.type) {
     case SAVE:
-      const key = [action.annotation_name, action.tile_overlay.id];
-      const old_results = state.results.getIn(key); 
-      return { results: state.results.setIn(key, action.tile_overlay) };
+      const key = ['results', action.annotation_name, action.tile_overlay.id];
+      const old_results = state.getIn(key); 
+      return state.setIn(key, action.tile_overlay); 
+    case UPDATE_SHOW_CELLS:
+    case UPDATE_SHOW_SUPERPIXELS:
+      return state.set(action.name, action.value);
     default: 
       return state;
   }

--- a/src/util/TileCellData.js
+++ b/src/util/TileCellData.js
@@ -2,9 +2,10 @@ import cv from 'opencv.js';
 const { List } = require("immutable");
 
 export default class TileCellData {
-  constructor(tile) {
-    this.id = tile.cacheKey;
-    this.tile = tile
+  constructor(superpixel_classification) {
+    this.id = superpixel_classification.tile.cacheKey;
+    this.tile = superpixel_classification.tile;
+    this.superpixel_classification = superpixel_classification;
 
     this.centroids_x = List();
     this.centroids_y = List();
@@ -13,16 +14,18 @@ export default class TileCellData {
     this.angles = List();
 
     this.canvas = document.createElement('canvas');
-    this.canvas.style.zIndex = "2";
+    this.canvas.style.zIndex = "3";
     this.context = this.canvas.getContext('2d');
-    this.canvas.width = tile.sourceBounds.width;
-    this.canvas.height = tile.sourceBounds.height;
-    this.canvas.id = this.id;
+    this.canvas.width = this.tile.sourceBounds.width;
+    this.canvas.height = this.tile.sourceBounds.height;
+    this.canvas.id = `tile_cell_data ${this.id}`;
+
+    this._segment();
   }
 
 
   copy() {
-    let c = new TileCellData(this.tile);
+    let c = new TileCellData(this.superpixel_classification.copy());
     c.centroids_x = this.centroids_x
     c.centroids_y = this.centroids_y
     c.widths = this.widths;
@@ -40,25 +43,102 @@ export default class TileCellData {
     this.angles = this.angles.push(a);
   }
 
-  segment() {
+  _segment2() {
     const rows = this.canvas.height;
     const cols = this.canvas.width;
-    let src = new cv.Mat(rows, cols, cv.CV_8UC4);
-    let dst = cv.Mat.zeros(rows, cols, cv.CV_8UC3);
+
+    let src = cv.Mat.zeros(rows, cols, cv.CV_8UC3);
     let mask = cv.Mat.zeros(rows, cols, cv.CV_8UC1);
-    // Iterate through every pixel
-    // pixel_classification is [1 = green, -1 = red, 0 = transparent]
     for (let i = 0; i < rows; i++) {
       for (let j = 0; j < cols; j++) {
         const index = i*cols + j;
-        if (this.pixel_classification[index] > 0) {
+        if (this.superpixel_classification.pixel_classification[index] > 0) {
           // classification 
           mask.ucharPtr(i, j)[0] = 255;
         }
-        src.ucharPtr(i, j)[0] = this.src.data[index*4 + 0]; // R
-        src.ucharPtr(i, j)[1] = this.src.data[index*4 + 1]; // G
-        src.ucharPtr(i, j)[2] = this.src.data[index*4 + 2]; // B
-        src.ucharPtr(i, j)[3] = this.src.data[index*4 + 3]; // A
+        src.ucharPtr(i, j)[0] = this.superpixel_classification.src.data[index*4 + 0]; // R
+        src.ucharPtr(i, j)[1] = this.superpixel_classification.src.data[index*4 + 1]; // G
+        src.ucharPtr(i, j)[2] = this.superpixel_classification.src.data[index*4 + 2]; // B
+      }
+    }
+    let opening = new cv.Mat();
+    let gray = new cv.Mat();
+    let coinsBg = new cv.Mat();
+    let coinsFg = new cv.Mat();
+    let distTrans = new cv.Mat();
+    let unknown = new cv.Mat();
+    let markers = new cv.Mat();
+
+    // get background
+    let M = cv.Mat.ones(3, 3, cv.CV_8U);
+    cv.erode(mask, gray, M);
+    cv.dilate(gray, opening, M);
+    cv.dilate(opening, coinsBg, M, new cv.Point(-1, -1), 3);
+    // distance transform
+    cv.distanceTransform(opening, distTrans, cv.DIST_L2, 5);
+    cv.normalize(distTrans, distTrans, 1, 0, cv.NORM_INF);
+    // get foreground
+    cv.threshold(distTrans, coinsFg, 0.7 * 1, 255, cv.THRESH_BINARY);
+    coinsFg.convertTo(coinsFg, cv.CV_8U, 1, 0);
+    cv.subtract(coinsBg, coinsFg, unknown);
+    // get connected components markers
+    cv.connectedComponents(coinsFg, markers);
+    for (let i = 0; i < markers.rows; i++) {
+        for (let j = 0; j < markers.cols; j++) {
+            markers.intPtr(i, j)[0] = markers.ucharPtr(i, j)[0] + 1;
+            if (unknown.ucharPtr(i, j)[0] == 255) {
+                markers.intPtr(i, j)[0] = 0;
+            }
+        }
+    }
+    cv.cvtColor(src, src, cv.COLOR_RGBA2RGB, 0);
+    cv.watershed(src, markers);
+
+    // draw barriers on mask
+    for (let i = 0; i < markers.rows; i++) {
+        for (let j = 0; j < markers.cols; j++) {
+            if (markers.intPtr(i, j)[0] == -1) {
+                mask.ucharPtr(i, j)[0] = 0; 
+            }
+        }
+    }
+    this.mask = mask;
+
+    let contours = new cv.MatVector();
+    let hierarchy = new cv.Mat();
+    cv.findContours(mask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
+
+    for (let i = 0; i < contours.size(); ++i) {
+      if (hierarchy.intPtr(0, i)[2] >= 0) continue;
+      //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
+      const cell_area = cv.contourArea(contours.get(i));
+
+      if (cell_area > 10) {
+        // fit an ellipse
+        const rotatedRect = cv.fitEllipse(contours.get(i));
+        // add ellipse to data
+        this.add_cell(rotatedRect.center.x, rotatedRect.center.y, rotatedRect.size.width, rotatedRect.size.height, rotatedRect.angle);
+      }
+    }
+  
+    src.delete(); gray.delete(); opening.delete(); coinsBg.delete();
+    coinsFg.delete(); distTrans.delete(); unknown.delete(); markers.delete(); M.delete();
+    contours.delete(); hierarchy.delete();
+  }
+
+  _segment() {
+    const rows = this.canvas.height;
+    const cols = this.canvas.width;
+
+    // write pixel classification to mask
+    let mask = cv.Mat.zeros(rows, cols, cv.CV_8UC1);
+    for (let i = 0; i < rows; i++) {
+      for (let j = 0; j < cols; j++) {
+        const index = i*cols + j;
+        if (this.superpixel_classification.pixel_classification[index] > 0) {
+          // classification 
+          mask.ucharPtr(i, j)[0] = 255;
+        }
       }
     }
 
@@ -68,54 +148,49 @@ export default class TileCellData {
     let max_area = 2*goal_max_area;
 
     //while (max_area > goal_max_area) {
-    for (let ii = 0; ii < 4; ++ii) {
-      cv.findContours(mask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
-      max_area = 0;
-      for (let i = 0; i < contours.size(); ++i) {
-        if (hierarchy.intPtr(0, i)[2] >= 0) continue;
-        const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
-        //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
-        const cell_area = cv.contourArea(contours.get(i));
-        if (cell_area > max_area) {
-          max_area = cell_area;
-        }
+    //for (let ii = 0; ii < 0; ++ii) {
+    //  cv.findContours(mask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
+    //  max_area = 0;
+    //  for (let i = 0; i < contours.size(); ++i) {
+    //    if (hierarchy.intPtr(0, i)[2] >= 0) continue;
+    //    const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
+    //    //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
+    //    const cell_area = cv.contourArea(contours.get(i));
+    //    if (cell_area > max_area) {
+    //      max_area = cell_area;
+    //    }
 
-        if (cell_area > goal_max_area) {
-          const rotatedRect = cv.fitEllipse(contours.get(i));
-          const vertices = cv.RotatedRect.points(rotatedRect);
-          const point1 = new cv.Point(0.5*(vertices[0].x + vertices[1].x), 0.5*(vertices[0].y + vertices[1].y));
-          const point2 = new cv.Point(0.5*(vertices[2].x + vertices[3].x), 0.5*(vertices[2].y + vertices[3].y));
-          const line_color = new cv.Scalar(0, 0, 0);
-          cv.line(mask, point1, point2, line_color, 2);
-        }
-      }
-      console.log(`iteration max area is ${max_area}`);
-    }
-    console.log(`final max area is ${max_area}`);
+    //    if (cell_area > goal_max_area) {
+    //      const rotatedRect = cv.fitEllipse(contours.get(i));
+    //      const vertices = cv.RotatedRect.points(rotatedRect);
+    //      const point1 = new cv.Point(0.5*(vertices[0].x + vertices[1].x), 0.5*(vertices[0].y + vertices[1].y));
+    //      const point2 = new cv.Point(0.5*(vertices[2].x + vertices[3].x), 0.5*(vertices[2].y + vertices[3].y));
+    //      const line_color = new cv.Scalar(0, 0, 0);
+    //      cv.line(mask, point1, point2, line_color, 2);
+    //    }
+    //  }
+    //  console.log(`iteration max area is ${max_area}`);
+    //}
+    //console.log(`final max area is ${max_area}`);
+    cv.findContours(mask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
 
     for (let i = 0; i < contours.size(); ++i) {
       if (hierarchy.intPtr(0, i)[2] >= 0) continue;
-      const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
       //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
       const cell_area = cv.contourArea(contours.get(i));
 
       if (cell_area > 10) {
+        // fit an ellipse
         const rotatedRect = cv.fitEllipse(contours.get(i));
-        cv.ellipse1(dst, rotatedRect, color, 1, cv.LINE_8);
+        // add ellipse to data
+        this.add_cell(rotatedRect.center.x, rotatedRect.center.y, rotatedRect.size.width, rotatedRect.size.height, rotatedRect.angle);
       }
     }
+  
+    console.log(`finished segmenting, found ${this.centroids_x.size} cells`);
 
-    // copy segmentation
-    for (let i = 0; i < dst.rows; i++) {
-        for (let j = 0; j < dst.cols; j++) {
-          const index = i*cols + j;
-          this.pixel_segmentation[index] = dst.ucharPtr(i, j)[0]; 
-        }
-    }
-
-    src.delete(); mask.delete(); dst.delete();
+    mask.delete();
     contours.delete(); hierarchy.delete();
-    this.segmented = true;
   }
 
   /// draws current classification data to the canvas
@@ -132,6 +207,8 @@ export default class TileCellData {
           angle: this.angles.get(i)
         };
 
+        //const color = new cv.Scalar(Math.ceil((i+1)/this.centroids_x.size * 255), 0, 0, 200);
+        const color = new cv.Scalar(0, 0, 0, 200);
         cv.ellipse1(dst, rotatedRect, color, 1, cv.LINE_8);
     }
 

--- a/src/util/TileCellData.js
+++ b/src/util/TileCellData.js
@@ -1,0 +1,155 @@
+import cv from 'opencv.js';
+const { List } = require("immutable");
+
+export default class TileCellData {
+  constructor(tile) {
+    this.id = tile.cacheKey;
+    this.tile = tile
+
+    this.centroids_x = List();
+    this.centroids_y = List();
+    this.widths = List();
+    this.heights = List();
+    this.angles = List();
+
+    this.canvas = document.createElement('canvas');
+    this.canvas.style.zIndex = "2";
+    this.context = this.canvas.getContext('2d');
+    this.canvas.width = tile.sourceBounds.width;
+    this.canvas.height = tile.sourceBounds.height;
+    this.canvas.id = this.id;
+  }
+
+
+  copy() {
+    let c = new TileCellData(this.tile);
+    c.centroids_x = this.centroids_x
+    c.centroids_y = this.centroids_y
+    c.widths = this.widths;
+    c.heights = this.heights;
+    c.angles = this.angles;
+    redraw();
+    return c;
+  }
+
+  add_cell(x, y, w, h, a) {
+    this.centroids_x = this.centroids_x.push(x);
+    this.centroids_y = this.centroids_y.push(y);
+    this.widths = this.widths.push(w);
+    this.heights = this.heights.push(h);
+    this.angles = this.angles.push(a);
+  }
+
+  segment() {
+    const rows = this.canvas.height;
+    const cols = this.canvas.width;
+    let src = new cv.Mat(rows, cols, cv.CV_8UC4);
+    let dst = cv.Mat.zeros(rows, cols, cv.CV_8UC3);
+    let mask = cv.Mat.zeros(rows, cols, cv.CV_8UC1);
+    // Iterate through every pixel
+    // pixel_classification is [1 = green, -1 = red, 0 = transparent]
+    for (let i = 0; i < rows; i++) {
+      for (let j = 0; j < cols; j++) {
+        const index = i*cols + j;
+        if (this.pixel_classification[index] > 0) {
+          // classification 
+          mask.ucharPtr(i, j)[0] = 255;
+        }
+        src.ucharPtr(i, j)[0] = this.src.data[index*4 + 0]; // R
+        src.ucharPtr(i, j)[1] = this.src.data[index*4 + 1]; // G
+        src.ucharPtr(i, j)[2] = this.src.data[index*4 + 2]; // B
+        src.ucharPtr(i, j)[3] = this.src.data[index*4 + 3]; // A
+      }
+    }
+
+    let contours = new cv.MatVector();
+    let hierarchy = new cv.Mat();
+    let goal_max_area = 200;
+    let max_area = 2*goal_max_area;
+
+    //while (max_area > goal_max_area) {
+    for (let ii = 0; ii < 4; ++ii) {
+      cv.findContours(mask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
+      max_area = 0;
+      for (let i = 0; i < contours.size(); ++i) {
+        if (hierarchy.intPtr(0, i)[2] >= 0) continue;
+        const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
+        //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
+        const cell_area = cv.contourArea(contours.get(i));
+        if (cell_area > max_area) {
+          max_area = cell_area;
+        }
+
+        if (cell_area > goal_max_area) {
+          const rotatedRect = cv.fitEllipse(contours.get(i));
+          const vertices = cv.RotatedRect.points(rotatedRect);
+          const point1 = new cv.Point(0.5*(vertices[0].x + vertices[1].x), 0.5*(vertices[0].y + vertices[1].y));
+          const point2 = new cv.Point(0.5*(vertices[2].x + vertices[3].x), 0.5*(vertices[2].y + vertices[3].y));
+          const line_color = new cv.Scalar(0, 0, 0);
+          cv.line(mask, point1, point2, line_color, 2);
+        }
+      }
+      console.log(`iteration max area is ${max_area}`);
+    }
+    console.log(`final max area is ${max_area}`);
+
+    for (let i = 0; i < contours.size(); ++i) {
+      if (hierarchy.intPtr(0, i)[2] >= 0) continue;
+      const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
+      //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
+      const cell_area = cv.contourArea(contours.get(i));
+
+      if (cell_area > 10) {
+        const rotatedRect = cv.fitEllipse(contours.get(i));
+        cv.ellipse1(dst, rotatedRect, color, 1, cv.LINE_8);
+      }
+    }
+
+    // copy segmentation
+    for (let i = 0; i < dst.rows; i++) {
+        for (let j = 0; j < dst.cols; j++) {
+          const index = i*cols + j;
+          this.pixel_segmentation[index] = dst.ucharPtr(i, j)[0]; 
+        }
+    }
+
+    src.delete(); mask.delete(); dst.delete();
+    contours.delete(); hierarchy.delete();
+    this.segmented = true;
+  }
+
+  /// draws current classification data to the canvas
+  redraw() {
+    
+    // draw ellipses using opencv
+    const rows = this.canvas.height;
+    const cols = this.canvas.width;
+    let dst = cv.Mat.zeros(rows, cols, cv.CV_8UC4);
+    for (let i = 0; i < this.centroids_x.size; ++i) {
+        const rotatedRect = {
+          center : new cv.Point(this.centroids_x.get(i), this.centroids_y.get(i)),
+          size : new cv.Size(this.widths.get(i), this.heights.get(i)),
+          angle: this.angles.get(i)
+        };
+
+        cv.ellipse1(dst, rotatedRect, color, 1, cv.LINE_8);
+    }
+
+    // copy opencv mat to imageData
+    const imageData = this.context.createImageData(this.canvas.width, this.canvas
+      .height);
+
+    for (let i = 0; i < dst.rows; i++) {
+      for (let j = 0; j < dst.cols; j++) {
+        const index = i*cols + j;
+        const nchannels = 4;
+        for (let c = 0; c < nchannels; c++) {
+          imageData.data[nchannels*index + c] = dst.ucharPtr(i, j)[c]; 
+        }
+      }
+    }
+
+    // Draw image data to the canvas
+    this.context.putImageData(imageData, 0, 0);
+  }
+}

--- a/src/util/TileOverlay.js
+++ b/src/util/TileOverlay.js
@@ -29,19 +29,15 @@ export default class TileOverlay {
     this.canvas.id = this.id;
     this.pixel_classification = new Int8Array(tile.sourceBounds.width * tile
       .sourceBounds.height);
-    this.pixel_segmentation = new Int8Array(tile.sourceBounds.width * tile
-      .sourceBounds.height);
     this.positive_superpixels = new Set();
     this.negative_superpixels = new Set();
     this.predict_superpixels = new Set();
-    this.segmented = false;
   }
 
 
   copy() {
     let c = new TileOverlay(this.tile, this.labels, this.features, this.src);
     c.pixel_classification = new Int8Array(this.pixel_classification);
-    c.pixel_segmentation = new Int8Array(this.pixel_segmentation);
     c.positive_superpixels = new Set(this.positive_superpixels);
     c.negative_superpixels = new Set(this.negative_superpixels);
     c.predict_superpixels = new Set(this.predict_superpixels);
@@ -139,83 +135,6 @@ export default class TileOverlay {
     return features;
   }
 
-  segment() {
-    const rows = this.canvas.height;
-    const cols = this.canvas.width;
-    let src = new cv.Mat(rows, cols, cv.CV_8UC4);
-    let dst = cv.Mat.zeros(rows, cols, cv.CV_8UC3);
-    let mask = cv.Mat.zeros(rows, cols, cv.CV_8UC1);
-    // Iterate through every pixel
-    // pixel_classification is [1 = green, -1 = red, 0 = transparent]
-    for (let i = 0; i < rows; i++) {
-      for (let j = 0; j < cols; j++) {
-        const index = i*cols + j;
-        if (this.pixel_classification[index] > 0) {
-          // classification 
-          mask.ucharPtr(i, j)[0] = 255;
-        }
-        src.ucharPtr(i, j)[0] = this.src.data[index*4 + 0]; // R
-        src.ucharPtr(i, j)[1] = this.src.data[index*4 + 1]; // G
-        src.ucharPtr(i, j)[2] = this.src.data[index*4 + 2]; // B
-        src.ucharPtr(i, j)[3] = this.src.data[index*4 + 3]; // A
-      }
-    }
-
-    let contours = new cv.MatVector();
-    let hierarchy = new cv.Mat();
-    let goal_max_area = 200;
-    let max_area = 2*goal_max_area;
-
-    //while (max_area > goal_max_area) {
-    for (let ii = 0; ii < 4; ++ii) {
-      cv.findContours(mask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
-      max_area = 0;
-      for (let i = 0; i < contours.size(); ++i) {
-        if (hierarchy.intPtr(0, i)[2] >= 0) continue;
-        const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
-        //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
-        const cell_area = cv.contourArea(contours.get(i));
-        if (cell_area > max_area) {
-          max_area = cell_area;
-        }
-
-        if (cell_area > goal_max_area) {
-          const rotatedRect = cv.fitEllipse(contours.get(i));
-          const vertices = cv.RotatedRect.points(rotatedRect);
-          const point1 = new cv.Point(0.5*(vertices[0].x + vertices[1].x), 0.5*(vertices[0].y + vertices[1].y));
-          const point2 = new cv.Point(0.5*(vertices[2].x + vertices[3].x), 0.5*(vertices[2].y + vertices[3].y));
-          const line_color = new cv.Scalar(0, 0, 0);
-          cv.line(mask, point1, point2, line_color, 2);
-        }
-      }
-      console.log(`iteration max area is ${max_area}`);
-    }
-    console.log(`final max area is ${max_area}`);
-
-    for (let i = 0; i < contours.size(); ++i) {
-      if (hierarchy.intPtr(0, i)[2] >= 0) continue;
-      const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
-      //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
-      const cell_area = cv.contourArea(contours.get(i));
-
-      if (cell_area > 10) {
-        const rotatedRect = cv.fitEllipse(contours.get(i));
-        cv.ellipse1(dst, rotatedRect, color, 1, cv.LINE_8);
-      }
-    }
-
-    // copy segmentation
-    for (let i = 0; i < dst.rows; i++) {
-        for (let j = 0; j < dst.cols; j++) {
-          const index = i*cols + j;
-          this.pixel_segmentation[index] = dst.ucharPtr(i, j)[0]; 
-        }
-    }
-
-    src.delete(); mask.delete(); dst.delete();
-    contours.delete(); hierarchy.delete();
-    this.segmented = true;
-  }
 
   /// draws current classification data to the canvas
   redraw() {
@@ -223,39 +142,20 @@ export default class TileOverlay {
     const imageData = this.context.createImageData(this.canvas.width, this.canvas
       .height);
 
-    if (!this.segmented) {
-      // pixel_classification is [1 = green, -1 = red, 0 = transparent]
-      for (let i = 0; i < imageData.width * imageData.height; i++) {
-        if (this.pixel_classification[i] > 0) {
-          // green
-          imageData.data[4 * i + 0] = 0;
-          imageData.data[4 * i + 1] = 255;
-          imageData.data[4 * i + 2] = 0;
-          imageData.data[4 * i + 3] = 100;
-        } else if (this.pixel_classification[i] < 0) {
-          // red 
-          imageData.data[4 * i + 0] = 255;
-          imageData.data[4 * i + 1] = 0;
-          imageData.data[4 * i + 2] = 0;
-          imageData.data[4 * i + 3] = 100;
-        }
-      }
-    } else {
-      // pixel_segmentation is [0 = background, X = cell number]
-      for (let i = 0; i < imageData.width * imageData.height; i++) {
-        if (this.pixel_segmentation[i] == 0) {
-          // transparent background
-          imageData.data[4 * i + 0] = 0;
-          imageData.data[4 * i + 1] = 0;
-          imageData.data[4 * i + 2] = 0;
-          imageData.data[4 * i + 3] = 0;
-        } else {
-          // color by region
-          imageData.data[4 * i + 0] = 0;
-          imageData.data[4 * i + 1] = this.pixel_segmentation[i];
-          imageData.data[4 * i + 2] = 0;
-          imageData.data[4 * i + 3] = 255;
-        }
+    // pixel_classification is [1 = green, -1 = red, 0 = transparent]
+    for (let i = 0; i < imageData.width * imageData.height; i++) {
+      if (this.pixel_classification[i] > 0) {
+        // green
+        imageData.data[4 * i + 0] = 0;
+        imageData.data[4 * i + 1] = 255;
+        imageData.data[4 * i + 2] = 0;
+        imageData.data[4 * i + 3] = 100;
+      } else if (this.pixel_classification[i] < 0) {
+        // red 
+        imageData.data[4 * i + 0] = 255;
+        imageData.data[4 * i + 1] = 0;
+        imageData.data[4 * i + 2] = 0;
+        imageData.data[4 * i + 3] = 100;
       }
     }
 

--- a/src/util/TileOverlay.js
+++ b/src/util/TileOverlay.js
@@ -37,6 +37,7 @@ export default class TileOverlay {
     this.segmented = false;
   }
 
+
   copy() {
     let c = new TileOverlay(this.tile, this.labels, this.features, this.src);
     c.pixel_classification = new Int8Array(this.pixel_classification);

--- a/src/util/TileOverlay.js
+++ b/src/util/TileOverlay.js
@@ -143,7 +143,8 @@ export default class TileOverlay {
     const rows = this.canvas.height;
     const cols = this.canvas.width;
     let src = new cv.Mat(rows, cols, cv.CV_8UC4);
-    let opening = cv.Mat.zeros(rows, cols, cv.CV_8UC1);
+    let dst = cv.Mat.zeros(rows, cols, cv.CV_8UC3);
+    let mask = cv.Mat.zeros(rows, cols, cv.CV_8UC1);
     // Iterate through every pixel
     // pixel_classification is [1 = green, -1 = red, 0 = transparent]
     for (let i = 0; i < rows; i++) {
@@ -151,7 +152,7 @@ export default class TileOverlay {
         const index = i*cols + j;
         if (this.pixel_classification[index] > 0) {
           // classification 
-          opening.ucharPtr(i, j)[0] = 255;
+          mask.ucharPtr(i, j)[0] = 255;
         }
         src.ucharPtr(i, j)[0] = this.src.data[index*4 + 0]; // R
         src.ucharPtr(i, j)[1] = this.src.data[index*4 + 1]; // G
@@ -160,43 +161,59 @@ export default class TileOverlay {
       }
     }
 
-    let background = new cv.Mat();
-    let foreground = new cv.Mat();
-    let distTrans = new cv.Mat();
-    let unknown = new cv.Mat();
-    let markers = new cv.Mat();
-    // get background
-    let M = cv.Mat.ones(3, 3, cv.CV_8U);
-    cv.dilate(opening, background, M, new cv.Point(-1, -1), 3);
-    // distance transform
-    cv.distanceTransform(opening, distTrans, cv.DIST_L2, 5);
-    cv.normalize(distTrans, distTrans, 1, 0, cv.NORM_INF);
-    // get foreground
-    cv.threshold(distTrans, foreground, 0.7 * 1, 255, cv.THRESH_BINARY);
-    foreground.convertTo(foreground, cv.CV_8U, 1, 0);
-    cv.subtract(background, foreground, unknown);
-    // get connected components markers
-    cv.connectedComponents(foreground, markers);
-    for (let i = 0; i < markers.rows; i++) {
-        for (let j = 0; j < markers.cols; j++) {
-            markers.intPtr(i, j)[0] = markers.ucharPtr(i, j)[0] + 1;
-            if (unknown.ucharPtr(i, j)[0] == 255) {
-                markers.intPtr(i, j)[0] = 0;
-            }
+    let contours = new cv.MatVector();
+    let hierarchy = new cv.Mat();
+    let goal_max_area = 200;
+    let max_area = 2*goal_max_area;
+
+    //while (max_area > goal_max_area) {
+    for (let ii = 0; ii < 4; ++ii) {
+      cv.findContours(mask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
+      max_area = 0;
+      for (let i = 0; i < contours.size(); ++i) {
+        if (hierarchy.intPtr(0, i)[2] >= 0) continue;
+        const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
+        //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
+        const cell_area = cv.contourArea(contours.get(i));
+        if (cell_area > max_area) {
+          max_area = cell_area;
         }
+
+        if (cell_area > goal_max_area) {
+          const rotatedRect = cv.fitEllipse(contours.get(i));
+          const vertices = cv.RotatedRect.points(rotatedRect);
+          const point1 = new cv.Point(0.5*(vertices[0].x + vertices[1].x), 0.5*(vertices[0].y + vertices[1].y));
+          const point2 = new cv.Point(0.5*(vertices[2].x + vertices[3].x), 0.5*(vertices[2].y + vertices[3].y));
+          const line_color = new cv.Scalar(0, 0, 0);
+          cv.line(mask, point1, point2, line_color, 2);
+        }
+      }
+      console.log(`iteration max area is ${max_area}`);
     }
-    //cv.cvtColor(src, src, cv.COLOR_RGBA2RGB, 0);
-    //cv.watershed(src, markers);
+    console.log(`final max area is ${max_area}`);
+
+    for (let i = 0; i < contours.size(); ++i) {
+      if (hierarchy.intPtr(0, i)[2] >= 0) continue;
+      const color = new cv.Scalar(Math.ceil((i+1)/contours.size() * 255), 0, 0);
+      //cv.drawContours(dst, contours, i, color, 1, cv.LINE_8, hierarchy, 100);
+      const cell_area = cv.contourArea(contours.get(i));
+
+      if (cell_area > 10) {
+        const rotatedRect = cv.fitEllipse(contours.get(i));
+        cv.ellipse1(dst, rotatedRect, color, 1, cv.LINE_8);
+      }
+    }
+
     // copy segmentation
-    for (let i = 0; i < markers.rows; i++) {
-        for (let j = 0; j < markers.cols; j++) {
+    for (let i = 0; i < dst.rows; i++) {
+        for (let j = 0; j < dst.cols; j++) {
           const index = i*cols + j;
-          this.pixel_segmentation[index] = foreground.intPtr(i, j)[0]; 
+          this.pixel_segmentation[index] = dst.ucharPtr(i, j)[0]; 
         }
     }
-    console.log(markers);
-    src.delete(); opening.delete(); background.delete();
-    foreground.delete(); distTrans.delete(); unknown.delete(); markers.delete(); M.delete();
+
+    src.delete(); mask.delete(); dst.delete();
+    contours.delete(); hierarchy.delete();
     this.segmented = true;
   }
 
@@ -224,15 +241,9 @@ export default class TileOverlay {
         }
       }
     } else {
-      // pixel_segmentation is [-1 = border, 1 = background, X = cell number]
+      // pixel_segmentation is [0 = background, X = cell number]
       for (let i = 0; i < imageData.width * imageData.height; i++) {
-        if (this.pixel_segmentation[i] < 1) {
-          // black borders
-          imageData.data[4 * i + 0] = 0;
-          imageData.data[4 * i + 1] = 0;
-          imageData.data[4 * i + 2] = 0;
-          imageData.data[4 * i + 3] = 100;
-        } else if (this.pixel_segmentation[i] == 1) {
+        if (this.pixel_segmentation[i] == 0) {
           // transparent background
           imageData.data[4 * i + 0] = 0;
           imageData.data[4 * i + 1] = 0;
@@ -243,7 +254,7 @@ export default class TileOverlay {
           imageData.data[4 * i + 0] = 0;
           imageData.data[4 * i + 1] = this.pixel_segmentation[i];
           imageData.data[4 * i + 2] = 0;
-          imageData.data[4 * i + 3] = 100;
+          imageData.data[4 * i + 3] = 255;
         }
       }
     }

--- a/src/util/TileOverlay.js
+++ b/src/util/TileOverlay.js
@@ -1,3 +1,4 @@
+import cv from 'opencv.js';
 
 export default class TileOverlay {
 //
@@ -12,13 +13,15 @@ export default class TileOverlay {
   ///  @param {Openseadragon.Tile} tile: image tile
   ///  @param {TypedArray} labels: maps pixels to superpixel number
   ///  @param {array} features: 26*nsuperpixels array of features
-  constructor(tile, labels, features) {
+  ///  @param {ImageData} src: the source image data for this tile
+  constructor(tile, labels, features, src) {
     this.id = tile.cacheKey;
     this.tile = tile;
+    this.src = src; 
     this.labels = new Int32Array(labels);
     this.nfeatures = 26;
     this.features = new Float64Array(features);
-    this.canvas = document.createElement("canvas");
+    this.canvas = document.createElement('canvas');
     this.canvas.style.zIndex = "2";
     this.context = this.canvas.getContext('2d');
     this.canvas.width = tile.sourceBounds.width;
@@ -26,17 +29,22 @@ export default class TileOverlay {
     this.canvas.id = this.id;
     this.pixel_classification = new Int8Array(tile.sourceBounds.width * tile
       .sourceBounds.height);
+    this.pixel_segmentation = new Int8Array(tile.sourceBounds.width * tile
+      .sourceBounds.height);
     this.positive_superpixels = new Set();
     this.negative_superpixels = new Set();
     this.predict_superpixels = new Set();
+    this.segmented = false;
   }
 
   copy() {
-    let c = new TileOverlay(this.tile, this.labels, this.features);
+    let c = new TileOverlay(this.tile, this.labels, this.features, this.src);
     c.pixel_classification = new Int8Array(this.pixel_classification);
+    c.pixel_segmentation = new Int8Array(this.pixel_segmentation);
     c.positive_superpixels = new Set(this.positive_superpixels);
     c.negative_superpixels = new Set(this.negative_superpixels);
     c.predict_superpixels = new Set(this.predict_superpixels);
+    c.segmented = this.segmented;
     c.redraw();
     return c;
   }
@@ -130,23 +138,112 @@ export default class TileOverlay {
     return features;
   }
 
+  segment() {
+    const rows = this.canvas.height;
+    const cols = this.canvas.width;
+    let src = new cv.Mat(rows, cols, cv.CV_8UC4);
+    let opening = cv.Mat.zeros(rows, cols, cv.CV_8UC1);
+    // Iterate through every pixel
+    // pixel_classification is [1 = green, -1 = red, 0 = transparent]
+    for (let i = 0; i < rows; i++) {
+      for (let j = 0; j < cols; j++) {
+        const index = i*cols + j;
+        if (this.pixel_classification[index] > 0) {
+          // classification 
+          opening.ucharPtr(i, j)[0] = 255;
+        }
+        src.ucharPtr(i, j)[0] = this.src.data[index*4 + 0]; // R
+        src.ucharPtr(i, j)[1] = this.src.data[index*4 + 1]; // G
+        src.ucharPtr(i, j)[2] = this.src.data[index*4 + 2]; // B
+        src.ucharPtr(i, j)[3] = this.src.data[index*4 + 3]; // A
+      }
+    }
+
+    let background = new cv.Mat();
+    let foreground = new cv.Mat();
+    let distTrans = new cv.Mat();
+    let unknown = new cv.Mat();
+    let markers = new cv.Mat();
+    // get background
+    let M = cv.Mat.ones(3, 3, cv.CV_8U);
+    cv.dilate(opening, background, M, new cv.Point(-1, -1), 3);
+    // distance transform
+    cv.distanceTransform(opening, distTrans, cv.DIST_L2, 5);
+    cv.normalize(distTrans, distTrans, 1, 0, cv.NORM_INF);
+    // get foreground
+    cv.threshold(distTrans, foreground, 0.7 * 1, 255, cv.THRESH_BINARY);
+    foreground.convertTo(foreground, cv.CV_8U, 1, 0);
+    cv.subtract(background, foreground, unknown);
+    // get connected components markers
+    cv.connectedComponents(foreground, markers);
+    for (let i = 0; i < markers.rows; i++) {
+        for (let j = 0; j < markers.cols; j++) {
+            markers.intPtr(i, j)[0] = markers.ucharPtr(i, j)[0] + 1;
+            if (unknown.ucharPtr(i, j)[0] == 255) {
+                markers.intPtr(i, j)[0] = 0;
+            }
+        }
+    }
+    //cv.cvtColor(src, src, cv.COLOR_RGBA2RGB, 0);
+    //cv.watershed(src, markers);
+    // copy segmentation
+    for (let i = 0; i < markers.rows; i++) {
+        for (let j = 0; j < markers.cols; j++) {
+          const index = i*cols + j;
+          this.pixel_segmentation[index] = foreground.intPtr(i, j)[0]; 
+        }
+    }
+    console.log(markers);
+    src.delete(); opening.delete(); background.delete();
+    foreground.delete(); distTrans.delete(); unknown.delete(); markers.delete(); M.delete();
+    this.segmented = true;
+  }
+
   /// draws current classification data to the canvas
   redraw() {
     // default for createImageData is transparent black
     const imageData = this.context.createImageData(this.canvas.width, this.canvas
       .height);
 
-    // Iterate through every pixel
-    // pixel_classification is [1 = green, -1 = red, 0 = transparent]
-    for (let i = 0; i < imageData.width * imageData.height; i++) {
-      if (this.pixel_classification[i] > 0) {
-        // green
-        imageData.data[4 * i + 1] = 255;
-        imageData.data[4 * i + 3] = 100;
-      } else if (this.pixel_classification[i] < 0) {
-        // red
-        imageData.data[4 * i + 0] = 255;
-        imageData.data[4 * i + 3] = 100;
+    if (!this.segmented) {
+      // pixel_classification is [1 = green, -1 = red, 0 = transparent]
+      for (let i = 0; i < imageData.width * imageData.height; i++) {
+        if (this.pixel_classification[i] > 0) {
+          // green
+          imageData.data[4 * i + 0] = 0;
+          imageData.data[4 * i + 1] = 255;
+          imageData.data[4 * i + 2] = 0;
+          imageData.data[4 * i + 3] = 100;
+        } else if (this.pixel_classification[i] < 0) {
+          // red 
+          imageData.data[4 * i + 0] = 255;
+          imageData.data[4 * i + 1] = 0;
+          imageData.data[4 * i + 2] = 0;
+          imageData.data[4 * i + 3] = 100;
+        }
+      }
+    } else {
+      // pixel_segmentation is [-1 = border, 1 = background, X = cell number]
+      for (let i = 0; i < imageData.width * imageData.height; i++) {
+        if (this.pixel_segmentation[i] < 1) {
+          // black borders
+          imageData.data[4 * i + 0] = 0;
+          imageData.data[4 * i + 1] = 0;
+          imageData.data[4 * i + 2] = 0;
+          imageData.data[4 * i + 3] = 100;
+        } else if (this.pixel_segmentation[i] == 1) {
+          // transparent background
+          imageData.data[4 * i + 0] = 0;
+          imageData.data[4 * i + 1] = 0;
+          imageData.data[4 * i + 2] = 0;
+          imageData.data[4 * i + 3] = 0;
+        } else {
+          // color by region
+          imageData.data[4 * i + 0] = 0;
+          imageData.data[4 * i + 1] = this.pixel_segmentation[i];
+          imageData.data[4 * i + 2] = 0;
+          imageData.data[4 * i + 3] = 100;
+        }
       }
     }
 

--- a/test/test_application_state.js
+++ b/test/test_application_state.js
@@ -1,6 +1,6 @@
 import chai from 'chai';
 import sinon from 'sinon';
-import DefaultMode from '../src/ApplicationState';
+import DefaultMode from '../src/components/ApplicationState';
 
 describe('DefaultMode', function() {
     it('Defaults to DisabledMode', function() {
@@ -33,7 +33,7 @@ describe('DisabledMode', function() {
         const sandbox = sinon.createSandbox();
         const openSpy = sandbox.spy();
         const menu = {
-            props: {
+            viewer: {
                 openseadragon: {
                     open: openSpy,
                 },

